types/joi/index.d.ts

```diff
@@ -966,7 +966,7 @@ export interface Rules<P extends object = any> {
     name: string;
     params?: ObjectSchema | {[key in keyof P]: SchemaLike; };
     setup?(this: ExtensionBoundSchema, params: P): Schema | void;
-    validate?<R = any>(this: ExtensionBoundSchema, params: P, value: any, state: State, options: ValidationOptions): Err | R;
+    validate?(this: ExtensionBoundSchema, params: P, value: any, state: State, options: ValidationOptions): Err | any;
```

Comment:
As mentioned in the PR comments, I don't think it's necessary to remove the type parameter here. You should be able to specify the type parameter manually, although as we discovered, TSC will no longer be able to infer the types of the arguments and you must also specify them manually.

If we remove the "<R = any>" default type parameter and just make it "<R>", I believe TSC can infer the return type from a rule's implementation of `validate()`.

I'd like to remove the `any` type because it means, by default, the return value is always `any`, because the union `Err | any` is the same as just `any`.

Reply:
You raise a good point, however both the before and after versions are incorrect.

We discourage the use of generic methods where the type parameter only appears in the return type, because typing `foo<string>()` is neither shorter nor clearer than typing `foo() as string`. 

On the other hand, as you pointed out @laurence-myers, `Err | any` is the same as `any`.

So this method should simply be 
`validate?(this: ExtensionBoundSchema, params: P, value: any, state: State, options: ValidationOptions): any;`

From a type checking perspective, `validate` can return anything. TypeScript has nothing to go on here to infer the return type. If you need to assert that you are expecting a specific type, the straightforward approach is to just use `as` on the return value.